New: Release script improvements

molant opened the pull request:
Draft PR for the release script.

Most of the logic is there except the part of publishing to npm and github, need to figure out how to prompt for the otp multiple times in a clean way (although hopefully we will be using a token). The rest should be working.

There are a few things I'm sure can be improved like the `skipReasons` in `main.ts`. There are quite a few that are repeated but I'm not sure on how to do it better (maybe `skipIfError` is always on when using `skipReasons`?).
I repeat a lot a pattern with observables that I'll probably refactor as well. In any case, I think this is a good moment to start reviewing it.

Another thig is that right now I'm commiting the changes package by package but it's really slow. I'm thinking about just doing one commit with a generic message `Update: Dependencies and package version`. That will save a few minutes. Thoughts?

Things pending:

* [x] npm publishing
* [x] publish changes on github (`git push`)
* [x] delete github token when using user/pass
* [x] test more the `private` and unpublished packages
* [x] test on a mac
* [x] remove dependencies from old script and old script
* [x] test `eslint-plugin-import`
* [x] add help menu
* [x] improve the go back state, maybe prompt if we want to revert or just revert if changes are discarded. The reason is that there might be an issue when publishing to npm or whe puhing to GitHub and we don't want to go back to the start state in those cases

You can give it a try locally, nothing is going to get commited into GitHub or npm. If you don't validate the changes it should revert back to the origin and you can always run `git reset --hard SHA` and you can find the SHA in the `release.log` file

molant commented:
BTW, the way code is organized is as follows:

`release/main.ts` --> The entry point. I'll probably change the the filename later
`release/tasks/*.ts` --> The tasks executed by `main.ts`
`release/lib/*.ts` --> Helpers and such

Probably best start with main and then go task by task to review this.

sarvaje commented on release/test.ts:
```diff
@@ -0,0 +1,54 @@
+            return Promise.resolve();
+        }
+    }, {
+        title: 'Step 3',
```
`Step 4`

molant replied:
This file needs to be deleted 🤦‍♂️

sarvaje commented on release/test.ts:
```diff
@@ -0,0 +1,54 @@
+        }
+    }]);
+
+tasks.run();
```
New line at the end of the file

antross commented:
> I'm thinking about just doing one commit with a generic message Update: Dependencies and package version. That will save a few minutes. Thoughts?

I think that's preferrable, including the `update-3rd-party` dependencies as well.

molant commented:
> I think that's preferrable, including the `update-3rd-party` dependencies as well.

Sounds good. I look into it although I might need it for rollback purposes. I'll keep you posted.

antross commented on release/lib/git-helpers.ts:
```diff
@@ -0,0 +1,302 @@
+        const { stdout } = await execa(command);
+        /**
+         * The format is `sha (Author date) lineContent`.E.g.:
+         * `c10092ade (Tony Ross 2018-08-22 17:39:49 -0500 77)   }`
```
I feel famous! 😎

antross commented on release/tasks/calculate-new-versions.ts:
```diff
@@ -0,0 +1,76 @@
+ *
+ * @param tag The tag of the commit
+ */
+const commitType = (tag: Tag): Bump => {
```
Maybe rename to something like `getBumpTypeForTag` to be a bit clearer?

antross commented on release/tasks/calculate-new-versions.ts:
```diff
@@ -0,0 +1,76 @@
+const calculatePackageNewVersion = (pkg: Package, bump: Bump): string => {
+
+    if (!pkg.publishedVersion) {
+        return pkg.content.newVersion;
```
Should this be covered in the types somewhere? `pkg.content` is currently `any` (which means `pkg.content.newVersion` is `any` as well).

molant replied:
This is most likely a typo. I think it should be `pkg.content.version`

antross commented on release/tasks/calculate-new-versions.ts:
```diff
@@ -0,0 +1,76 @@
+    // Step 2
+    updateDependencies(ctx);
+
+    // Step 3
```
Probably also worth repeating what each step does here (even though it's in the comment for the whole function). It took me a second to get my bearings on the first read.

antross commented on release/@types/custom.ts:
```diff
@@ -0,0 +1,68 @@
+    user?: string;
+    pass?: string;
+    otp?: string;
+    token?: string;
```
Probably better to split this into two separate sub-types as well and union them together in `GitHubAuth`.

antross commented on release/lib/update-3rd-party.ts:
```diff
@@ -0,0 +1,60 @@
+};
+
+export const updateThirdPartyResources = () => {
+    return new Observable((observer) => {
```
What does this do?

antross commented on release/tasks/save-changes.ts:
```diff
@@ -0,0 +1,15 @@
+import { Context } from '../@types/custom';
+
+/** Saves in disk the changes to `package.json` of all packages */
+export const saveChanges = (ctx: Context) => {
```
Maybe rename to `savePackageChanges` to be a bit more specific?

antross commented on release/tasks/validate-changes.ts:
```diff
@@ -0,0 +1,46 @@
+        const newVersion = pkg.content.version;
+
+        if (newVersion !== oldVersion) {
+            console.log(`${packageName + new Array(maxLength + 1 - packageName.length).join(' ')} ${oldVersion} -> ${newVersion}`);
```
```suggestion
            console.log(`${packageName.padEnd(maxLength + 1, ' ')} ${oldVersion} -> ${newVersion}`);
```

antross commented on release/tasks/run-tests.ts:
```diff
@@ -0,0 +1,86 @@
+        return;
+    }
+
+    // npm run lint && npm run build && npm run test-only
```
Why not run `yarn test` which we already configure in each package to run all three steps?

molant replied:
That was my intention with line 13 but forgot to update the rest of the code 😅

molant commented:
> test `eslint-plugin-import` 

It doesn't seem to do what we need. I've looked at some of the resolvers for this package but none do what we need. Creating a resolver doesn't seem overcomplicated so I'm thinking about creating one that will find the first `package.json` in the tree for each file and then use the `dependencies` name to check if it is installed or not.

molant commented:
@antross, @sarvaje I'm going to run a full run with `--dry-run` to see how it goes and without publishing to GitHub but you should start looking into this when you have the time 🙏

Example of the help menu:
![image](https://user-images.githubusercontent.com/606594/54788576-b7d40000-4bec-11e9-846a-6363cb608f68.png)

Example of the task menu:
![image](https://user-images.githubusercontent.com/606594/54788595-d2a67480-4bec-11e9-8bd7-75790f9ac990.png)

antross commented:
Per our discussion, don't forget to re-add the checks to verify we only release from `master` on the official `webhintio/hint` repo.

antross commented:
@molant The `yarn release` command still points to the old, non-existent release script in `package.json`.

molant commented on release/lib/git-helpers.ts:
```diff
@@ -0,0 +1,310 @@
+ * Returns the GitHub user information associated with a give commit.
+ * @param commitSHA The sha of the commit to get the information from.
+ */
+const getCommitAuthorInfo = (commitSHA: string): Promise<Author | null> => {
```
I originally had it with `async/await` but because the info for the commits is run in parallel I was be querying the same information multiple times and reach the GitHub limit for requests and everything crashed. Now that everything is stored as promises we only request the bare minimum.

molant commented on release/tasks/run-tests.ts:
```diff
@@ -0,0 +1,105 @@
+                 */
+                await buildPackages(pkg.references, packages);
+            } catch (e) {
+                observer.error(`Build failed for package ${pkg.name} dependencies`);
```
I think the `observer.error`  stops the execution but now I'm not sure :/

molant replied:
Yup, [`error` means completion](http://reactivex.io/documentation/contract.html) from what I read:

> An Observable may make zero or more OnNext notifications, each representing a single emitted item, and it may then follow those emission notifications by either an **OnCompleted or an OnError notification, but not both.**

molant commented:
Feedback should be addressed and conflicts resolved!

molant commented:
Rebased with master and fixed the last round of issues. Now if something breaks in the middle we will continue executing tasks and ask for clean up at the end. Also capturing exception with `tasks.run().catch()` and telling the user to look the logs to get more information.

Also tested with mac and everything seems to be working here as well. I can't repro the node version limitation in this platform so maybe check the OS before validating the version? @antross what do you think?

antross commented on release/lib/utils.ts:
```diff
@@ -0,0 +1,181 @@
+/**
+ * The output stream to `release.log`
+ */
+// TODO: We should wait for the 'open' event to be sure everything is fine
```
Was this something you wanted to do before merging?

molant replied:
not really, it complicates the code quite a bit and we take long enough to have the stream ready before the first use... I'll just leave the comment here so at least we know I was aware in case something happens.

antross commented on release/tasks/get-packages.ts:
```diff
@@ -0,0 +1,116 @@
+        });
+
+        const pkgPaths = await globby([
+            'packages/**/*/package.json',
```
Do we support nested packages generally, or would `'packages/*/package.json'` suffice here?

molant replied:
I don't know what the future holds and it was easy enough to have it this way and just negate the `fixtures`

molant commented:
@antross I've published the refactor we talked about but I'm going to change `request` with `octokit/rest` for the GitHub interaction. The reason is that we are still being throttled from time to time and I'd like to avoid that so this is still not ready to be merged.

molant commented:
I've migrated to octokit and not getting any more issues requesting info. Code should be simpler now too. @sarvaje want to take another look?  You were the one suggesting to migrate to it initially 😊

sarvaje commented on release/lib/git-helpers.ts:
```diff
@@ -1,78 +1,85 @@
+        warn: debug
+    },
+    throttle: {
+        onAbuseLimit: (retryAfter: number, options: { method: string; url: string; request: { retryCount: number } }) => {
```
We should have this in the `online-service` :)

sarvaje commented on release/lib/git-helpers.ts:
```diff
@@ -203,6 +210,34 @@ const extractDataFromCommit = (sha: string): Promise<Commit> => {
+
+const createOctokitFromUserPass = (auth: GitHubAuth) => {
+    if ('token' in auth) {
+        return new Client();
```
shouldn't this be `return createOctokitFromToken(auth.token)` ?

molant replied:
That code should never be hit, it was there so TypeScript doesn't complain about possible missing properties and ESLint about not having consistent returns. I'll add it for completeness with a comment.
